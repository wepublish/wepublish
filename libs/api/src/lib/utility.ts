import {MemberPlan, PaymentMethod} from '@prisma/client'
import formatISO from 'date-fns/formatISO'
import {GraphQLFieldResolver, GraphQLIsTypeOfFn, GraphQLObjectType} from 'graphql'
import {delegateToSchema, ExecutionResult, IDelegateToSchemaOptions, Transform} from 'graphql-tools'
import {Context} from './context'
import {TeaserStyle} from './db/block'
import {SettingRestriction} from '@wepublish/settings/api'
import {SubscriptionWithRelations} from './db/subscription'
import {UserWithRelations} from './db/user'
import {InvalidSettingValueError} from './error'

export const MAX_PAYLOAD_SIZE = '1MB'

export const ONE_HOUR_IN_MILLISECONDS = 60 * 60 * 1000
export const ONE_DAY_IN_MILLISECONDS = 24 * ONE_HOUR_IN_MILLISECONDS
export const FIFTEEN_MINUTES_IN_MILLISECONDS = 900000
export const ONE_MONTH_IN_MILLISECONDS = 31 * ONE_DAY_IN_MILLISECONDS

export const USER_PROPERTY_LAST_LOGIN_LINK_SEND = '_wepLastLoginLinkSentTimestamp'

export function mapSubscriptionsAsCsv(
  subscriptions: (SubscriptionWithRelations & {
    user: UserWithRelations
    paymentMethod: PaymentMethod
    memberPlan: MemberPlan
  })[]
) {
  let csvStr =
    [
      'id',
      'firstName',
      'name',
      'preferredName',
      'email',
      'active',
      'createdAt',
      'modifiedAt',

      'company',
      'streetAddress',
      'streetAddress2',
      'zipCode',
      'city',
      'country',

      'memberPlan',
      'memberPlanID',
      'paymentPeriodicity',
      'monthlyAmount',
      'autoRenew',
      'startsAt',
      'paidUntil',
      'paymentMethod',
      'paymentMethodID',
      'deactivationDate',
      'deactivationReason'
    ].join(',') + '\n'

  for (const subscription of subscriptions) {
    const user = subscription?.user
    const memberPlan = subscription?.memberPlan
    const paymentMethod = subscription?.paymentMethod
    // if (!user) continue
    csvStr +=
      [
        user?.id,
        `${sanitizeCsvContent(user?.firstName)}`,
        `${sanitizeCsvContent(user?.name)}`,
        `${sanitizeCsvContent(user?.preferredName)}`,
        `${user?.email ?? ''}`,
        user?.active,
        user?.createdAt ? formatISO(user.createdAt, {representation: 'date'}) : '',
        user?.modifiedAt ? formatISO(user.modifiedAt, {representation: 'date'}) : '',
        `${sanitizeCsvContent(user?.address?.company)}`,
        `${sanitizeCsvContent(user?.address?.streetAddress)}`,
        `${sanitizeCsvContent(user?.address?.streetAddress2)}`,
        `${sanitizeCsvContent(user?.address?.zipCode)}`,
        `${sanitizeCsvContent(user?.address?.city)}`,
        `${sanitizeCsvContent(user?.address?.country)}`,
        sanitizeCsvContent(memberPlan?.name),
        subscription?.memberPlanID ?? '',
        subscription?.paymentPeriodicity ?? '',
        subscription?.monthlyAmount ?? '',
        subscription?.autoRenew ?? '',
        subscription?.startsAt ? formatISO(subscription.startsAt, {representation: 'date'}) : '',
        subscription?.paidUntil
          ? formatISO(subscription.paidUntil, {representation: 'date'})
          : 'no pay',
        sanitizeCsvContent(paymentMethod?.name),
        subscription?.paymentMethodID ?? '',
        subscription?.deactivation
          ? formatISO(subscription.deactivation.date, {representation: 'date'})
          : '',
        subscription?.deactivation?.reason ?? ''
      ].join(',') + '\r\n'
  }

  return csvStr
}

/**
 * according to rfc 4180
 * https://www.ietf.org/rfc/rfc4180.txt
 * @param input
 */
function sanitizeCsvContent(input: string | undefined | null) {
  // according rfc 4180 2.7.
  const escapeDoubleQuotes = (input || '').toString().replace(/[#"]/g, '""')
  // according rfc 4180 2.5. / 2.6.
  return `"${escapeDoubleQuotes}"`
}

// https://gist.github.com/mathewbyrne/1280286#gistcomment-2588056
export function slugify(str: string) {
  return str
    .toLowerCase()
    .trim()
    .replace(/[ÀÁÂÃÄÅÆĀĂĄẠẢẤẦẨẪẬẮẰẲẴẶ]/gi, 'a')
    .replace(/[ÇĆĈČ]/gi, 'c')
    .replace(/[ÐĎĐÞ]/gi, 'd')
    .replace(/[ÈÉÊËĒĔĖĘĚẸẺẼẾỀỂỄỆ]/gi, 'e')
    .replace(/[ĜĞĢǴ]/gi, 'g')
    .replace(/[ĤḦ]/gi, 'h')
    .replace(/[ÌÍÎÏĨĪĮİỈỊ]/gi, 'i')
    .replace(/[Ĵ]/gi, 'j')
    .replace(/[Ĳ]/gi, 'ij')
    .replace(/[Ķ]/gi, 'k')
    .replace(/[ĹĻĽŁ]/gi, 'l')
    .replace(/[Ḿ]/gi, 'm')
    .replace(/[ÑŃŅŇ]/gi, 'n')
    .replace(/[ÒÓÔÕÖØŌŎŐỌỎỐỒỔỖỘỚỜỞỠỢǪǬƠ]/gi, 'o')
    .replace(/[Œ]/gi, 'oe')
    .replace(/[ṕ]/gi, 'p')
    .replace(/[ŔŖŘ]/gi, 'r')
    .replace(/[ßŚŜŞŠ]/gi, 's')
    .replace(/[ŢŤ]/gi, 't')
    .replace(/[ÙÚÛÜŨŪŬŮŰŲỤỦỨỪỬỮỰƯ]/gi, 'u')
    .replace(/[ẂŴẀẄ]/gi, 'w')
    .replace(/[ẍ]/gi, 'x')
    .replace(/[ÝŶŸỲỴỶỸ]/gi, 'y')
    .replace(/[ŹŻŽ]/gi, 'z')
    .replace(/[·/_,:;\\']/gi, '-')
    .replace(/\s+/g, '-')
    .replace(/[^\w-]+/g, '')
    .replace(/--+/g, '-')
    .replace(/^-+/, '')
    .replace(/-+$/, '')
}

export function base64Encode(str: string): string {
  return Buffer.from(str).toString('base64')
}

export function base64Decode(str: string): string {
  return Buffer.from(str, 'base64').toString()
}

export const IsProxiedSymbol = Symbol('isProxied')

export function markResultAsProxied(result: any) {
  if (!result) return null

  for (const key in result) {
    const value = result[key]

    if (typeof value === 'object' && value != null) {
      markResultAsProxied(value)
    }
  }

  return Object.assign(result, {[IsProxiedSymbol]: true})
}

export function isSourceProxied<T>(source: T): source is T & {__typename?: string} {
  return (source as any)[IsProxiedSymbol] == true
}

export function createProxyingResolver<TSource, TContext, TArgs = {[argName: string]: any}>(
  resolver: GraphQLFieldResolver<TSource, TContext, TArgs>
): GraphQLFieldResolver<TSource, TContext, TArgs> {
  return (source, args, context, info) => {
    if (isSourceProxied(source)) {
      return (source as any)[info.path.key]
    }

    return resolver(source, args, context, info)
  }
}

export function createProxyingIsTypeOf<TSource, TContext>(
  isTypeOf: GraphQLIsTypeOfFn<TSource, TContext>
): GraphQLIsTypeOfFn<TSource, TContext> {
  return function (this: GraphQLObjectType, source, context, info) {
    return isSourceProxied(source)
      ? source.__typename === this.name
      : isTypeOf(source, context, info)
  }
}

export function mapEnumsBack(result: any) {
  if (!result) return null

  for (const key in result) {
    const value = result[key]

    if (typeof value === 'object' && value !== null) {
      mapEnumsBack(value)
    }
  }
  if (
    result.__typename === 'ArticleTeaser' ||
    result.__typename === 'PeerArticleTeaser' ||
    result.__typename === 'PageTeaser' ||
    result.__typename === 'EventTeaser' ||
    result.__typename === 'CustomTeaser'
  ) {
    switch (result.style) {
      case 'DEFAULT':
        return Object.assign(result, {style: TeaserStyle.Default})
      case 'LIGHT':
        return Object.assign(result, {style: TeaserStyle.Light})
      case 'TEXT':
        return Object.assign(result, {style: TeaserStyle.Text})
    }
  }
  return result
}

class ResetGraphQLEnums implements Transform {
  transformResult(result: ExecutionResult) {
    // FIXME: WPC-415 created
    return mapEnumsBack(result)
  }
}

export async function delegateToPeerSchema(
  peerID: string,
  fetchAdminEndpoint: boolean,
  context: Context,
  opts: Omit<IDelegateToSchemaOptions, 'schema'>
) {
  const schema = fetchAdminEndpoint
    ? await context.loaders.peerAdminSchema.load(peerID)
    : await context.loaders.peerSchema.load(peerID)

  if (!schema) return null

  return markResultAsProxied(
    await delegateToSchema({
      ...opts,
      schema,
      transforms: [new ResetGraphQLEnums(), ...(opts.transforms ?? [])]
    })
  )
}

export function capitalizeFirstLetter(str: string): string {
  return str.charAt(0).toUpperCase() + str.slice(1)
}

export function countRichtextChars(blocksCharLength: number, nodes: any) {
  return nodes.reduce((charLength: number, node: any) => {
    if (!node.text && !node.children) return charLength
    if (node.text) {
      return charLength + (node.text as string).length
    }
    return countRichtextChars(charLength, node.children)
  }, blocksCharLength)
}

export type DiscriminateUnion<T, K extends keyof T, V extends T[K]> = T extends Record<K, V>
  ? T
  : never

export type MapDiscriminatedUnion<T extends Record<K, string>, K extends keyof T> = {
  [V in T[K]]: DiscriminateUnion<T, K, V>
}

export function isObject<T>(unknown: unknown): unknown is Record<any, T> {
  return typeof unknown === 'object' && unknown != null && !Array.isArray(unknown)
}

export function isArray<T>(unknown: unknown): unknown is T[] {
  return Array.isArray(unknown)
}

export function isString(unknown: unknown): unknown is string {
  return typeof unknown === 'string'
}

export function isBoolean(unknown: unknown): unknown is boolean {
  return typeof unknown === 'boolean'
}

export function checkSettingRestrictions(
  val: unknown,
  currentVal: unknown,
  restriction: SettingRestriction | undefined
) {
  if (!restriction) {
    return
  }

  if (typeof val !== typeof currentVal) {
    throw new InvalidSettingValueError()
  }

  if (restriction.allowedValues?.boolChoice && typeof val !== 'boolean') {
    throw new InvalidSettingValueError()
  }

  if (typeof val === 'number') {
    if (restriction.maxValue && val > restriction.maxValue) {
      throw new InvalidSettingValueError()
    }

    if (restriction.minValue && val < restriction.minValue) {
      throw new InvalidSettingValueError()
    }
  }

  if (typeof val === 'string') {
    if (restriction.inputLength && val.length > restriction.inputLength) {
      throw new InvalidSettingValueError()
    }

    if (
      restriction.allowedValues?.stringChoice &&
      !restriction.allowedValues.stringChoice.includes(val)
    ) {
      throw new InvalidSettingValueError()
    }
  }
}
