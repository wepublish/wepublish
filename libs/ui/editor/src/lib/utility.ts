import {PaymentPeriodicity, SortOrder} from '@wepublish/editor/api'
import {DocumentNode, OperationDefinitionNode} from 'graphql'
import {de, enUS, fr} from 'date-fns/locale'
import nanoid from 'nanoid'
import {useCallback, useEffect, useMemo, useRef, useState} from 'react'

import {ApolloClient, ApolloLink, createHttpLink, InMemoryCache} from '@apollo/client'

export const addOrUpdateOneInArray = (
  array: Record<string | 'id', any>[] | null | undefined,
  entry: Record<string | 'id', any>
) => {
  let isNew = true

  if (!array) {
    return [entry]
  }
  const updated = array.map(item => {
    if (item.id !== entry.id) {
      // This isn't the item we care about - keep it as-is
      return item
    }
    isNew = false
    // Otherwise, this is the one we want - return an updated value
    return {
      ...item,
      ...entry
    }
  })

  if (isNew) {
    return [...updated, entry]
  }

  return updated
}

export function generateID(): string {
  return nanoid()
}

// https://gist.github.com/mathewbyrne/1280286#gistcomment-2588056
export function slugify(str: string) {
  return str
    .toLowerCase()
    .trim()
    .replace(/[ÀÁÂÃÅÆĀĂĄẠẢẤẦẨẪẬẮẰẲẴẶ]/gi, 'a')
    .replace(/[Ä]/gi, 'ae')

    .replace(/[ÇĆĈČ]/gi, 'c')
    .replace(/[ÐĎĐÞ]/gi, 'd')
    .replace(/[ÈÉÊËĒĔĖĘĚẸẺẼẾỀỂỄỆ]/gi, 'e')
    .replace(/[ĜĞĢǴ]/gi, 'g')
    .replace(/[ĤḦ]/gi, 'h')
    .replace(/[ÌÍÎÏĨĪĮİỈỊ]/gi, 'i')
    .replace(/[Ĵ]/gi, 'j')
    .replace(/[Ĳ]/gi, 'ij')
    .replace(/[Ķ]/gi, 'k')
    .replace(/[ĹĻĽŁ]/gi, 'l')
    .replace(/[Ḿ]/gi, 'm')
    .replace(/[ÑŃŅŇ]/gi, 'n')
    .replace(/[ÒÓÔÕØŌŎŐỌỎỐỒỔỖỘỚỜỞỠỢǪǬƠ]/gi, 'o')
    .replace(/[ŒÖ]/gi, 'oe')

    .replace(/[ṕ]/gi, 'p')
    .replace(/[ŔŖŘ]/gi, 'r')
    .replace(/[ŚŜŞŠ]/gi, 's')
    .replace(/[ß]/gi, 'ss')
    .replace(/[ŢŤ]/gi, 't')
    .replace(/[ÙÚÛŨŪŬŮŰŲỤỦỨỪỬỮỰƯ]/gi, 'u')
    .replace(/[Ü]/gi, 'ue')

    .replace(/[ẂŴẀẄ]/gi, 'w')
    .replace(/[ẍ]/gi, 'x')
    .replace(/[ÝŶŸỲỴỶỸ]/gi, 'y')
    .replace(/[ŹŻŽ]/gi, 'z')
    .replace(/[·/_,:;\\']/gi, '-')
    .replace(/\s+/g, '-')
    .replace(/[^\w\-]+/g, '') //eslint-disable-line
    .replace(/--+/g, '-')
    .replace(/^-+/, '')
    .replace(/-+$/, '')
}

export function useScript(src: string, checkIfLoaded: () => boolean, crossOrigin = false) {
  const scriptRef = useRef<HTMLScriptElement | null>(null)

  const [isLoading, setLoading] = useState(false)
  const [isLoaded, setLoaded] = useState(() => checkIfLoaded())

  useEffect(() => {
    if (isLoading && !isLoaded && !scriptRef.current) {
      const script = document.createElement('script')

      script.src = src
      script.async = true
      script.defer = true
      script.onload = () => setLoaded(true)
      script.crossOrigin = crossOrigin ? 'anonymous' : null

      document.head.appendChild(script)
      scriptRef.current = script
    }
  }, [isLoading, crossOrigin, src, isLoaded])

  const load = useCallback(() => {
    setLoading(true)
  }, [])

  const result = useMemo(
    () => ({
      isLoading,
      isLoaded,
      load
    }),
    [isLoading, isLoaded, load]
  )

  if (typeof window !== 'object') {
    return {
      isLoaded: false,
      isLoading: false,
      load: () => {
        /* do nothing */
      }
    }
  }

  return result
}

export function getOperationNameFromDocument(node: DocumentNode) {
  const firstOperation = node.definitions.find(
    node => node.kind === 'OperationDefinition'
  ) as OperationDefinitionNode

  if (!firstOperation?.name?.value) throw new Error("Coulnd't find operation name.")
  return firstOperation.name.value
}

export function transformCssStringToObject(styleCustom: string): Record<string, unknown> {
  const styleRules = styleCustom.split(';')
  if (styleRules.length === 0) return {}
  return styleRules.reduce((previousValue: Record<string, unknown>, currentValue: string) => {
    const [key, value] = currentValue.split(':')
    if (key && value) {
      return Object.assign(previousValue, {[key.trim()]: value.trim()})
    }
    return previousValue
  }, {})
}

export type SortType = 'asc' | 'desc' | null
export function mapTableSortTypeToGraphQLSortOrder(sortType: SortType): SortOrder | null {
  switch (sortType) {
    case 'desc':
      return SortOrder.Descending
    case 'asc':
      return SortOrder.Ascending
    default:
      return null
  }
}

export const DEFAULT_TABLE_PAGE_SIZES = [10, 20, 50, 100]
export const DEFAULT_TABLE_IMAGE_PAGE_SIZES = [5, 10, 15]
export const DEFAULT_MAX_TABLE_PAGES = 5

export const ALL_PAYMENT_PERIODICITIES: PaymentPeriodicity[] = [
  PaymentPeriodicity.Monthly,
  PaymentPeriodicity.Quarterly,
  PaymentPeriodicity.Biannual,
  PaymentPeriodicity.Yearly
]

export enum StateColor {
  pending = '#f8def2',
  published = '#e1f8de',
  draft = '#f8efde',
  none = 'white'
}

export function validateURL(url: string) {
  if (url) {
    const pattern = new RegExp(
      '^(https?:\\/\\/)?' + // protocol
        '((([a-z\\d]([a-z\\d-]*[a-z\\d])*)\\.)+[a-z]{2,}|' + // domain name
        '((\\d{1,3}\\.){3}\\d{1,3}))' + // OR ip (v4) address
        '(\\:\\d+)?(\\/[-a-z\\d%_.~+]*)*' + // port and path
        '(\\?[;&a-z\\d%_.~+=-]*)?' + // query string
        '(\\#[-a-z\\d_]*)?$',
      'i'
    )
    return pattern.test(url)
  }
  return false
}

export function flattenDOMTokenList(list: DOMTokenList) {
  let string = ''
  for (const element of Array.from(list)) {
    string = `${string} ${element}`
  }
  return string.substring(1)
}

export interface ClientSettings {
  readonly apiURL: string
  readonly peerByDefault: boolean
  readonly imgMinSizeToCompress: number
}

export enum ElementID {
  Settings = 'settings',
  ReactRoot = 'react-root'
}

export function getSettings(): ClientSettings {
  const defaultSettings = {
    apiURL: 'http://localhost:4000',
    peerByDefault: false,
    imgMinSizeToCompress: 10
  }

  const settingsJson = document.getElementById(ElementID.Settings)

  return settingsJson
    ? JSON.parse(document.getElementById(ElementID.Settings)!.textContent!)
    : defaultSettings
}

export interface ClientSettings {
  readonly apiURL: string
  readonly peerByDefault: boolean
  readonly imgMinSizeToCompress: number
}

export enum LocalStorageKey {
  SessionToken = 'sessionToken',
  ImageListLayout = 'imageListLayout'
}

const authLink = new ApolloLink((operation, forward) => {
  const token = localStorage.getItem(LocalStorageKey.SessionToken)
  const context = operation.getContext()

  operation.setContext({
    headers: {
      authorization: token ? `Bearer ${token}` : '',
      ...context.headers
    },
    credentials: 'include',
    ...context
  })

  return forward(operation)
})

export function getApiClientV2() {
  const {apiURL} = getSettings()
  return new ApolloClient({
    link: authLink.concat(createHttpLink({uri: `${apiURL}/v2`, fetch})),
    cache: new InMemoryCache()
  })
}

/**
 * Helper function to read env variable IMG_MIN_SIZE_TO_COMPRESS
 */
export function getImgMinSizeToCompress(): number {
  const {imgMinSizeToCompress} = getSettings()

  return imgMinSizeToCompress
}

export type UnionToIntersection<U> = (U extends any ? (k: U) => void : never) extends (
  k: infer I
) => void
  ? I
  : never

export type ValueConstructor<T> = T | (() => T)

export function isValueConstructor<T>(value: T | (() => T)): value is () => T {
  return typeof value === 'function'
}

export function isFunctionalUpdate<T>(value: React.SetStateAction<T>): value is (value: T) => T {
  return typeof value === 'function'
}

export const AVAILABLE_LANG = [
  {id: 'en', lang: 'en_US', name: 'English', locale: enUS},
  {id: 'fr', lang: 'fr_FR', name: 'Français', locale: fr},
  {id: 'de', lang: 'de_CH', name: 'Deutsch', locale: de}
]
