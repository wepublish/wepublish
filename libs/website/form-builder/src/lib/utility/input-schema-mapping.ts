import {ComponentType} from 'react'
import {ControllerFieldState, ControllerRenderProps} from 'react-hook-form'
import * as v from 'valibot'

export type InputComponentProps = {
  fieldState: ControllerFieldState
  field: ControllerRenderProps
  schema: v.BaseSchema<any, any, any>
  name: string
  title: string | null
  description: string | null
}

export type InputSchemaMapping = readonly [
  v.BaseSchema<any, any, any>,
  ComponentType<InputComponentProps>
]

export type FormSchemaMapping = ReadonlyArray<InputSchemaMapping>
