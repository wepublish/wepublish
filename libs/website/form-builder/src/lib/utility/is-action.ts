import * as v from 'valibot';

export const isBrandAction = (
  input: unknown
): input is v.BrandAction<any, any> => {
  return (input as any).type === 'brand';
};

export const hasBrandAction = (input: unknown): boolean => {
  if (isBrandAction(input)) {
    return input.name;
  }

  if (!isPipeAction(input)) {
    return false;
  }

  const results = input.pipe.find(schem => {
    return hasBrandAction(schem);
  });

  return !!results;
};

export const isPipeAction = (
  input: unknown
): input is v.SchemaWithPipe<
  [
    v.BaseSchema<unknown, unknown, v.BaseIssue<unknown>>,
    v.PipeItem<unknown, unknown, v.BaseIssue<unknown>>,
  ]
> => {
  return 'pipe' in (input as any);
};

export const isDescriptionAction = (
  input: unknown
): input is v.DescriptionAction<unknown, string> => {
  return (input as any).type === 'description';
};

export const isTitleAction = (
  input: unknown
): input is v.TitleAction<unknown, string> => {
  return (input as any).type === 'title';
};
