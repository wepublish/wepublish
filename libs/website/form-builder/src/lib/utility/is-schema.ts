import * as v from 'valibot';
import { InputSchemaMapping } from './input-schema-mapping';

export const isArraySchema = (
  input: InputSchemaMapping[0]
): input is v.ArraySchema<any, any> => {
  return input.type === 'array';
};

export const isObjectSchema = (
  input: InputSchemaMapping[0]
): input is v.ObjectSchema<any, any> => {
  return input.type === 'object';
};

export const isEnumSchema = (
  input: InputSchemaMapping[0]
): input is v.EnumSchema<any, any> => {
  return input.type === 'enum';
};

export const isBooleanSchema = (
  input: InputSchemaMapping[0]
): input is v.BooleanSchema<any> => {
  return input.type === 'boolean';
};

export const isStringSchema = (
  input: InputSchemaMapping[0]
): input is v.StringSchema<any> => {
  return input.type === 'string';
};
