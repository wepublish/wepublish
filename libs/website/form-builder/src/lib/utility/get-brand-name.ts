import {isBrandAction, isPipeAction} from './is-action'

export const getBrandName = (input: unknown) => {
  if (isBrandAction(input)) {
    return input.name
  }

  if (!isPipeAction(input)) {
    return null
  }

  const results = input.pipe.flatMap(schem => {
    const res = getBrandName(schem)

    return res ? [res] : []
  }) as string[]

  return results.at(-1)
}
