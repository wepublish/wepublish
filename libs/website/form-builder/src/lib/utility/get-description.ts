import {isDescriptionAction, isPipeAction, isTitleAction} from './is-action'

export const getDescription = (input: unknown) => {
  if (isDescriptionAction(input)) {
    return input.description
  }

  if (!isPipeAction(input)) {
    return null
  }

  const results = input.pipe.flatMap(schem => {
    const res = getDescription(schem)

    return res ? [res] : []
  }) as string[]

  return results.at(-1)
}

export const getTitle = (input: unknown) => {
  if (isTitleAction(input)) {
    return input.title
  }

  if (!isPipeAction(input)) {
    return null
  }

  const results = input.pipe.flatMap(schem => {
    const res = getTitle(schem)

    return res ? [res] : []
  }) as string[]

  return results.at(-1)
}
