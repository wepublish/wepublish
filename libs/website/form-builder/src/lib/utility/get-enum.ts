import { InputSchemaMapping } from './input-schema-mapping';
import { isEnumSchema } from './is-schema';

export const getEnums = (
  input: InputSchemaMapping[0]
): Record<string, string> | null => {
  if (isEnumSchema(input)) {
    return input.enum;
  }

  return null;
};
