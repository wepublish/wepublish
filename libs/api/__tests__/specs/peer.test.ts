import {ApolloServer} from 'apollo-server-express'
import {FetchMock} from 'jest-fetch-mock'
import fetch from 'node-fetch'
import {
  CreatePeer,
  CreatePeerInput,
  DeletePeer,
  Peer,
  PeerList,
  PeerProfile,
  PeerProfileInput,
  UpdatePeer,
  UpdatePeerProfile
} from '../api/private'
import fakePeerAdminSchema from '../fakePeerAdminSchema.json'

import {createGraphQLTestClientWithPrisma, generateRandomString} from '../utility'
;(fetch as unknown as FetchMock).mockResponse(JSON.stringify(fakePeerAdminSchema))

let testServerPrivate: ApolloServer

beforeAll(async () => {
  try {
    const setupClient = await createGraphQLTestClientWithPrisma()
    testServerPrivate = setupClient.testServerPrivate
  } catch (error) {
    console.log('Error', error)

    throw new Error('Error during test setup')
  }
})

describe('Peers', () => {
  describe('can be created/edited/deleted:', () => {
    const ids: string[] = []
    beforeAll(async () => {
      const input: CreatePeerInput = {
        name: `Peer Test ${ids.length}`,
        slug: generateRandomString(),
        hostURL: 'https://host-url.ch/',
        token: `token${ids.length}`
      }
      const res = await testServerPrivate.executeOperation({
        query: CreatePeer,
        variables: {
          input
        }
      })
      ids.unshift(res.data?.createPeer?.id)
    })

    test('can be created', async () => {
      const input: CreatePeerInput = {
        name: 'Create Peer Test',
        slug: generateRandomString(),
        hostURL: 'https://host-url.ch/',
        token: 'testTokenABC123'
      }
      const res = await testServerPrivate.executeOperation({
        query: CreatePeer,
        variables: {
          input
        }
      })
      expect(res).toMatchSnapshot({
        data: {
          createPeer: {
            id: expect.any(String),
            slug: expect.any(String)
          }
        }
      })
      ids.unshift(res.data?.createPeer?.id)
    })

    test('can be read in list', async () => {
      const res = await testServerPrivate.executeOperation({
        query: PeerList
      })

      expect(res.data?.peers).not.toHaveLength(0)
    })

    test('can be read by id', async () => {
      const res = await testServerPrivate.executeOperation({
        query: Peer,
        variables: {
          id: ids[0]
        }
      })
      expect(res).toMatchSnapshot({
        data: {
          peer: {
            id: expect.any(String),
            slug: expect.any(String)
          }
        }
      })
    })

    test('can be updated', async () => {
      const res = await testServerPrivate.executeOperation({
        query: UpdatePeer,
        variables: {
          input: {
            name: 'Updated Peer',
            slug: generateRandomString(),
            hostURL: 'https://new-host-url.ch/',
            token: 'newTestTokenABC123'
          },
          id: ids[0]
        }
      })

      expect(res).toMatchSnapshot({
        data: {
          updatePeer: {
            id: expect.any(String),
            slug: expect.any(String)
          }
        }
      })
    })

    test('can be deleted', async () => {
      const res = await testServerPrivate.executeOperation({
        query: DeletePeer,
        variables: {
          id: ids[0]
        }
      })

      expect(res).toMatchSnapshot({
        data: {
          deletePeer: expect.any(Object)
        }
      })
      expect(res.data?.deletePeer.id).toBe(ids[0])

      ids.shift()
    })

    test('can read peer profile', async () => {
      const res = await testServerPrivate.executeOperation({
        query: PeerProfile
      })
      expect(res).toMatchSnapshot()
    })

    test('can update peer profile', async () => {
      const input: PeerProfileInput = {
        name: 'test peer profile',
        themeColor: '#4287f5',
        themeFontColor: '#d67c15',
        callToActionText: [{text: 'rich text call to action'}],
        callToActionURL: 'calltoactionurl.ch/'
      }

      const res = await testServerPrivate.executeOperation({
        query: UpdatePeerProfile,
        variables: {
          input
        }
      })

      expect(res).toMatchSnapshot()
    })
  })
})
