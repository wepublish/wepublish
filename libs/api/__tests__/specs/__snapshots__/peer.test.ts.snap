// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Peers can be created/edited/deleted: can be created 1`] = `
{
  "data": {
    "createPeer": {
      "hostURL": "https://host-url.ch/",
      "id": Any<String>,
      "isDisabled": false,
      "name": "Create Peer Test",
      "slug": Any<String>,
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": {
    "headers": Headers {
      Symbol(map): {},
    },
  },
}
`;

exports[`Peers can be created/edited/deleted: can be deleted 1`] = `
{
  "data": {
    "deletePeer": Any<Object>,
  },
  "errors": undefined,
  "extensions": undefined,
  "http": {
    "headers": Headers {
      Symbol(map): {},
    },
  },
}
`;

exports[`Peers can be created/edited/deleted: can be read by id 1`] = `
{
  "data": {
    "peer": {
      "hostURL": "https://host-url.ch/",
      "id": Any<String>,
      "isDisabled": false,
      "name": "Create Peer Test",
      "slug": Any<String>,
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": {
    "headers": Headers {
      Symbol(map): {},
    },
  },
}
`;

exports[`Peers can be created/edited/deleted: can be updated 1`] = `
{
  "data": {
    "updatePeer": {
      "hostURL": "https://new-host-url.ch/",
      "id": Any<String>,
      "isDisabled": false,
      "name": "Updated Peer",
      "slug": Any<String>,
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": {
    "headers": Headers {
      Symbol(map): {},
    },
  },
}
`;

exports[`Peers can be created/edited/deleted: can read peer profile 1`] = `
{
  "data": {
    "peerProfile": {
      "callToActionImage": null,
      "callToActionImageURL": "",
      "callToActionText": [],
      "callToActionURL": "",
      "hostURL": "https://fakeURL",
      "logo": null,
      "name": "",
      "themeColor": "#000000",
      "themeFontColor": "#ffffff",
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": {
    "headers": Headers {
      Symbol(map): {},
    },
  },
}
`;

exports[`Peers can be created/edited/deleted: can update peer profile 1`] = `
{
  "data": {
    "updatePeerProfile": {
      "callToActionImage": null,
      "callToActionImageURL": null,
      "callToActionText": [
        {
          "text": "rich text call to action",
        },
      ],
      "callToActionURL": "calltoactionurl.ch/",
      "hostURL": "https://fakeURL",
      "logo": null,
      "name": "test peer profile",
      "themeColor": "#4287f5",
      "themeFontColor": "#d67c15",
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": {
    "headers": Headers {
      Symbol(map): {},
    },
  },
}
`;
